raft-tech/TANF-app
# Add GitHub Action to trigger deploys on PR label assignment [720]

### Summary of Changes
Enables manually triggered deploys through CircleCI by label assignment on GitHub PRs.

* Adds a GitHub Action workflow that runs on the `labeled` event for Pull Requests in order to make a cURL call to the Circle CI API to trigger a `deploy` job. This workflow only executes if the tag name is explicitly `Deploy with CircleCI` and if all status checks are passing in CI for the PR.
* Updates the CircleCI config to create a `DOCKER_TAG` environment variable which is the branch name with the literal character `/` replaced with `-`. This variable is then used to define the name of the image tag within DockerHub that is used for deployments. This change was necessary because the `/` is an invalid character for docker tags and we were previously only using the branch name to define tags - since many of our branch names include the `/` character this would be rejected for those branches without the change.
* Updates README file to include how to manually trigger deployments and the process for setting that up on other repositories.

Addresses https://github.com/raft-tech/TANF-app/issues/720

### How to Test
* Assign the `Deploy with CircleCI` tag to this PR (if already applied, remove it first to trigger another deploy)
* Visit the [GitHub Actions](https://github.com/raft-tech/TANF-app/actions) tab and ensure that a new deploy job was started
* Click in to the latest execution of the workflow and visit the link shown in `build_url` as seen below:

![circleci_job_url](https://user-images.githubusercontent.com/22626085/110541379-a8119700-80f5-11eb-8e3e-df9554115978.png)
* Check that the deployment completes and shows in the PR as seen below:

![Screenshot from 2021-03-11 16-04-05](https://user-images.githubusercontent.com/22626085/110855585-90b4e400-8284-11eb-9e25-0accd51b6583.png)

* If the PR status checks are failing or incomplete the deployment will be skipped as shown below:

![circleci_action_skipped_state](https://user-images.githubusercontent.com/22626085/110855635-a1fdf080-8284-11eb-9de2-6bd33c97831a.png)

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

  - [x] Contributors with `triage` or higher access to the repo can trigger a deploy to `Development` on demand from a GitHub PR by applying the label `Deploy with CircleCI`
  - [x] README is updated, to explain how to trigger a deploy on demand
  - [x] Documented process for communicating deploys to the team
  - [ ] Stretch goal: Find a way to link the PR in the CircleCI deploy history so we can see what is on the dev site at the moment

As facilitator/product manager, @kniz-raft will decide if ACs are met from Raft's perspective.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

N/A Infrastructure changes

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*


### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

N/A no frontend changes

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*


### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*


### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*
